### DISCUSSION TITLE
Issue with uploaded S3 images after moving from 2.11 to 3.1

### DISCUSSION THREAD

**Discussion by MrFehr** [solidusio/solidus#4499] [Troubleshooting] [answered]
Hi all,

I have recently upgrade from 2.11 to 3.1, and unfortunately found out after pushing to prod that none of the previously uploaded images work. I have had to temporarily hide them on the live to allow it to still be usable, but it is a bigger issue internally as we cannot manage any orders due to /edit erroring when also trying to find the images.

The main error that occurs is: `ActionView::Template::Error (undefined method 'processed' for nil:NilClass):` whenever I try to load any page using an image.

Beforehand on 2.11 my setup was done the following:

```
if Rails.env.production?
    attachment_config = {
      s3_credentials: {
        bucket:            ENV.fetch('S3_BUCKET_NAME'),
        access_key_id:     ENV.fetch('AWS_ACCESS_KEY_ID'),
        secret_access_key: ENV.fetch('AWS_SECRET_ACCESS_KEY')
      },

      storage:        :s3,
      s3_headers:     { 'Cache-Control' => 'max-age=31557600' },
      s3_protocol:    'https',
      s3_region:      ENV.fetch('AWS_REGION'),
      bucket:         ENV.fetch('S3_BUCKET_NAME'),
      url:            ':s3_domain_url',

      styles: {
        mini:    '128x128>',
        small:   '256x256>',
        product: '512x512>',
        large:   '1024x1024>'
      },

      path:          '/:class/:id/:style/:basename.:extension',
      default_url:   'noimage/:style.png',
      default_style: 'product',
    }

    attachment_config.each do |key, value|
      Spree::Image.attachment_definitions[:attachment][key.to_sym] = value
    end
end
```

This unfortunately won't deploy to Heroku due to an error `NoMethodError: undefined method 'has_one_attached' for #<Class:0x0000556c9d23c800>` so I've had to remove it.

I found @kennyadsl recommend a [solution to this problem](https://github.com/solidusio/solidus/discussions/4030) which I've added + deleted my `attachment_config`. This allows a deploy, but I once again get the `undefined method 'processed'` error again.

I have also tried:

- Adding `kt-paperclip` + `image_processing` to `Gemfile.rb`
- Adding `config.image_attachment_module = 'Spree::Image::ActiveStorageAttachment'` and `config.taxon_attachment_module = 'Spree::Taxon::ActiveStorageAttachment'` to `spree.rb` as per https://guides.solidus.io/developers/products-and-variants/product-images
- Adding + removing the `attachment_config` as per for `aws.rb` https://github.com/solidusio/solidus/wiki/Storing-images-on-S3-and-CDN's
- Tried this alternative version https://gist.github.com/ArunMichaelDsouza/00029719b65028ab7e1a
- Adding AWS config to storage.yml

I am really not sure where to go from here, but it is a pressing issue that I really need some help with. It's probably something small but I am not greatly familiar with this side of the codebase so I am currently in the dark as to what I'm actually doing.

Essentially, what I need is our old 2.11 images to be accessible and usable on this new 3.1 version. MANY thanks in advance! 🙌🏻

**Comment by kennyadsl** [comment, marked as answer]
Did you add 

```ruby
config.image_attachment_module = 'Spree::Image::PaperclipAttachment'
config.taxon_attachment_module = 'Spree::Taxon::PaperclipAttachment'
```

to the `spree.rb`? By default, Solidus it's trying to use ActiveStorage now and you have to instruct it to behave as before.

**Reply by MrFehr** [reply]
I did not. This has fixed it all! 🥳 one day I am gonna fly to Pescara and give you a big kiss my man 😘😂 thanks a lot! ❤️

Are you able to advise something for me please?

I have these three pieces of code relating to image size:

```
  # Change default size values of uploaded images
  config.product_image_styles = {
    mini: '128x128>',
    small: '256x256>',
    product: '512x512>',
    large: '1024x1024>'
  }
```

```
  if Rails.env.production?
    attachment_config = {
      s3_credentials: {
        bucket:            ENV["S3_BUCKET_NAME"],
        access_key_id:     ENV["AWS_ACCESS_KEY_ID"],
        secret_access_key: ENV["AWS_SECRET_ACCESS_KEY"]
      },

      storage:        :s3,
      s3_headers:     { "Cache-Control" => "max-age=31557600" },
      s3_protocol:    "https",
      s3_region:      ENV["AWS_REGION"],
      bucket:         ENV["S3_BUCKET_NAME"],
      url:            ":s3_domain_url",

      styles: {
        mini:    "128x128>",
        small:   "256x256>",
        product: "512x512>",
        large:   "1024x1024>"
      },

      path:          "/:class/:id/:style/:basename.:extension",
      default_url:   "noimage/:style.png",
      default_style: "product",
    }

    attachment_config.each do |key, value|
      Spree::Image.attachment_definitions[:attachment][key.to_sym] = value
    end
  end
```

```
# frozen_string_literal: true

module TyresDirect::ImageModifier
  private

  def prepended(klass)
    klass.attachment_definitions[:attachment][:styles] = {
      mini: '128x128>',
      small: '256x256>',
      product: '512x512>',
      large: '1024x1024>'
    }
  end

  Spree::Image.prepend self
end
```

What one(s) do I actually need? Also is there anything I need to be careful of going forward with new uploads?

**Reply by kennyadsl** [reply]
With Solidus 3.1, I think you can just keep the first definition of `styles` and remove the others.
